File: app/src/main/java/aromko/de/wishlist/viewModel/WishlistViewModel.java
package aromko.de.wishlist.viewModel;


import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import aromko.de.wishlist.database.FirebaseQueryLiveData;
import aromko.de.wishlist.model.Wishlist;
import aromko.de.wishlist.tasks.AppExecutors;

public class WishlistViewModel extends ViewModel {

    private static final String DB_PATH_WISHLISTS = "wishLists";
    private static final DatabaseReference LISTS_REF = FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS);

    private final FirebaseQueryLiveData liveData = new FirebaseQueryLiveData(LISTS_REF);
    private final MediatorLiveData<List<Wishlist>> listsLiveData = new MediatorLiveData<>();

    private FirebaseUser fFirebaseUser = FirebaseAuth.getInstance().getCurrentUser();

    public WishlistViewModel() {
        listsLiveData.addSource(liveData, new Observer<DataSnapshot>() {
            @Override
            public void onChanged(@Nullable final DataSnapshot dataSnapshot) {
                if (dataSnapshot != null) {
                    final List<Wishlist> lists = new ArrayList<>();
                    new AppExecutors().mainThread().execute(new Runnable() {
                        @Override
                        public void run() {
                            for (final DataSnapshot snapshot : dataSnapshot.getChildren()) {
                                Wishlist wishlist = snapshot.getValue(Wishlist.class);
                                wishlist.setKey(snapshot.getKey().toString());
                                String currentUid = fFirebaseUser.getUid();
                                if (wishlist.getAllowedUsers() != null && wishlist.getAllowedUsers().containsKey(currentUid) && wishlist.getAllowedUsers().get(currentUid).equals(true)) {
                                    lists.add(wishlist);
                                }
                            }
                            listsLiveData.postValue(lists);
                        }
                    });

                } else {
                    listsLiveData.setValue(null);
                }
            }
        });
    }

    public String insertList(String text, boolean isFavortieList) {
        String wishlistId = FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS).push().getKey();
        Map<String, Object> allowedUser = new HashMap<>();
        allowedUser.put(fFirebaseUser.getUid(), true);
        Wishlist wishlist = new Wishlist(text, System.currentTimeMillis() / 1000, allowedUser, 0, isFavortieList);
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).setValue(wishlist);
        return wishlistId;
    }

    @NonNull
    public LiveData<List<Wishlist>> getListsLiveData() {
        return listsLiveData;
    }

    public void addUserToWishlist(String wishlistId) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist currentWishlist = dataSnapshot.getValue(Wishlist.class);
                Map<String, Object> allowedUsers = new HashMap<>();
                if (currentWishlist.getAllowedUsers() != null) {
                    allowedUsers.putAll(currentWishlist.getAllowedUsers());
                }
                allowedUsers.put(fFirebaseUser.getUid(), true);
                currentWishlist.setAllowedUsers(allowedUsers);
                dataSnapshot.getRef().setValue(currentWishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void updateList(String wishlistId, final String name) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist currentWishlist = dataSnapshot.getValue(Wishlist.class);
                currentWishlist.setName(name);
                dataSnapshot.getRef().setValue(currentWishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void deleteList(String wishlistId) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist currentWishlist = dataSnapshot.getValue(Wishlist.class);
                currentWishlist.getAllowedUsers().remove(fFirebaseUser.getUid());
                dataSnapshot.getRef().setValue(currentWishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }
}

File: app/src/main/java/aromko/de/wishlist/viewModel/WishlistViewModel.kt
package aromko.de.wishlist.viewModel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import aromko.de.wishlist.database.FirebaseQueryLiveData
import aromko.de.wishlist.model.Wishlist
import aromko.de.wishlist.tasks.AppExecutors
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.*

class WishlistViewModel : ViewModel() {
    private val liveData = FirebaseQueryLiveData(LISTS_REF)
    val listsLiveData = MediatorLiveData<List<Wishlist?>?>()
    private val fFirebaseUser = FirebaseAuth.getInstance().currentUser
    fun insertList(text: String?, isFavortieList: Boolean): String? {
        val wishlistId = FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS").push().key
        val allowedUser: MutableMap<String?, Any?> = HashMap()
        allowedUser[fFirebaseUser!!.uid] = true
        val wishlist = Wishlist(text, System.currentTimeMillis() / 1000, allowedUser, 0, isFavortieList)
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId").setValue(wishlist)
        return wishlistId
    }

    fun getListsLiveData(): LiveData<List<Wishlist?>?> {
        return listsLiveData
    }

    fun addUserToWishlist(wishlistId: String?) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val currentWishlist = dataSnapshot.getValue(Wishlist::class.java)
                val allowedUsers: MutableMap<String?, Any?> = HashMap()
                if (currentWishlist?.allowedUsers != null) {
                    allowedUsers.putAll(currentWishlist.allowedUsers!!)
                }
                allowedUsers[fFirebaseUser!!.uid] = true
                currentWishlist?.allowedUsers = allowedUsers
                dataSnapshot.ref.setValue(currentWishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun updateList(wishlistId: String?, name: String?) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val currentWishlist = dataSnapshot.getValue(Wishlist::class.java)
                currentWishlist?.name = name
                dataSnapshot.ref.setValue(currentWishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun deleteList(wishlistId: String?) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val currentWishlist = dataSnapshot.getValue(Wishlist::class.java)
                currentWishlist?.allowedUsers?.remove(fFirebaseUser!!.uid)
                dataSnapshot.ref.setValue(currentWishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    companion object {
        private const val DB_PATH_WISHLISTS = "wishLists"
        private val LISTS_REF = FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS")
    }

    init {
        listsLiveData.addSource(liveData) { dataSnapshot: DataSnapshot? ->
            if (dataSnapshot != null) {
                val lists: MutableList<Wishlist?> = ArrayList()
                AppExecutors().mainThread().execute {
                    for (snapshot in dataSnapshot.children) {
                        val wishlist = snapshot.getValue(Wishlist::class.java)
                        wishlist?.key = snapshot.key
                        val currentUid = fFirebaseUser!!.uid
                        if (wishlist?.allowedUsers != null && wishlist.allowedUsers!!.containsKey(currentUid) && wishlist.allowedUsers!![currentUid] == true) {
                            lists.add(wishlist)
                        }
                    }
                    listsLiveData.postValue(lists)
                }
            } else {
                listsLiveData.setValue(null)
            }
        }
    }
}